daphne_worker/src/storage_proxy/metrics.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2024 Cloudflare, Inc. All rights reserved.
+// SPDX-License-Identifier: BSD-3-Clause
+
+use std::sync::OnceLock;
+
+use daphne::auth::BearerToken;
+use prometheus::{register_int_counter_vec, Encoder, IntCounterVec, TextEncoder};
+use url::Url;
+use worker::Env;
+
+struct Config {
+    prometheus_server: Url,
+
+    bearer_token: BearerToken,
+}
+
+impl Config {
+    fn from_env(env: &Env) -> Option<&'static Self> {
+        const DAP_METRICS_PUSH_SERVER_URL: &str = "DAP_METRICS_PUSH_SERVER_URL";
+        const DAP_METRICS_PUSH_BEARER_TOKEN: &str = "DAP_METRICS_PUSH_BEARER_TOKEN";
+        static CONFIG: OnceLock<Option<Config>> = OnceLock::new();
+
+        CONFIG
+            .get_or_init(|| {
+                match (
+                    env.var(DAP_METRICS_PUSH_SERVER_URL),
+                    env.var(DAP_METRICS_PUSH_BEARER_TOKEN),
+                ) {
+                    (Ok(server_str), Ok(bearer_token_str)) => {
+                        let prometheus_server = match server_str.to_string().parse() {
+                            Ok(server) => server,
+                            Err(error) => {
+                                tracing::error!(error = ?error, "invalid server url");
+                                return None;
+                            }
+                        };
+                        Some(Self {
+                            prometheus_server,
+                            bearer_token: BearerToken::from(bearer_token_str.to_string()),
+                        })
+                    }
+                    (Err(..), Err(..)) => None,
+                    (Ok(..), Err(..)) => {
+                        tracing::error!("failed to configure metrics push: missing bearer token");
+                        None
+                    }
+                    (Err(..), Ok(..)) => {
+                        tracing::error!("failed to configure metrics push: missing server URL");
+                        None
+                    }
+                }
+            })
+            .as_ref()
+    }
+}
+
+pub struct Metrics {
```

Comment:
For consistency with the `daphne_server` crate? 
```suggestion
pub struct StorageProxyMetrics {
```

Reply:
it's already inside of the `storage_proxy` module, maybe it doesn't need that qualification

Reply:
Up to you, just a suggestion.